gms/gossiper.cc

```diff
@@ -2150,27 +2211,104 @@ future<> gossiper::wait_for_gossip(std::chrono::milliseconds initial_delay, std:
 }
 
 future<> gossiper::wait_for_gossip_to_settle() {
-    static constexpr std::chrono::milliseconds GOSSIP_SETTLE_MIN_WAIT_MS{5000};
-    auto force_after = _cfg.skip_wait_for_gossip_to_settle();
+    logger.info("Waiting for gossip to settle with gossip query...");
     auto do_enable_features = [this] {
         return async([this] {
             if (!std::exchange(_gossip_settled, true)) {
                maybe_enable_features();
             }
         });
     };
+    auto force_after = _cfg.skip_wait_for_gossip_to_settle();
     if (force_after == 0) {
         return do_enable_features();
     }
-    return wait_for_gossip(GOSSIP_SETTLE_MIN_WAIT_MS, force_after).then([this, do_enable_features] {
+    return wait_for_gossip_with_query().then([do_enable_features] {
         return do_enable_features();
+    }).handle_exception([this, do_enable_features] (std::exception_ptr ep) {
+        logger.debug("Failed to run wait_for_gossip_with_query failed: {}", ep);
+        static constexpr std::chrono::milliseconds GOSSIP_SETTLE_MIN_WAIT_MS{5000};
+        logger.info("Waiting for gossip to settle without gossip query...");
+        return wait_for_gossip(GOSSIP_SETTLE_MIN_WAIT_MS).then([this, do_enable_features] {
+            return do_enable_features();
+        });
     });
 }
 
 future<> gossiper::wait_for_range_setup() {
-    logger.info("Waiting for pending range setup...");
-    auto ring_delay = std::chrono::milliseconds(_cfg.ring_delay_ms());
-    return wait_for_gossip(ring_delay);
+    logger.info("Waiting for pending range setup with gossip query...");
+    return gossiper::wait_for_gossip_with_query().handle_exception([this] (std::exception_ptr ep) {
+        logger.debug("Failed to run wait_for_gossip_with_query failed: {}", ep);
+        auto ring_delay = service::get_local_storage_service().get_ring_delay();
+        logger.info("Waiting for pending range setup without gossip query...");
+        return wait_for_gossip(ring_delay);
+    });
+}
+
+// Wait for gossip to settle with a rpc verb to query status of remote nodes
+// directly. If a node does not support the rpc verb or a node is down, we
+// will fall back to the old method, i.e., wait_for_gossip().
+future<> gossiper::wait_for_gossip_with_query() {
```

Comment:
We should separate the **data structures** we're managing (i.e. token ring status in this case) from the **method of management**.

Gossip is one way of managing/updating the structures, based on anti-entropy and eventual consistency.

However, the new method you're introducing in this patchset **is not gossip**. It's a different way of checking whether the data structures are synchronized, not based on gossip but based on immediate synchronization. Therefore I believe it should lie outside of `gossiper`.

Reply:
It is true it is not only about gossip, but it is highly related to gossip. We query the status of gossip and tokens in token_metadata. 

Where do you suggest we put this into?

Reply:
There are two steps in the propagation. The first step is the exchange of information (endpoint states) which is done in the gossip layer. The second step is building/updating internal data structures from that, on all the shards. The synchronization algorithm should not know about higher layers which derive representation from that basic information. Changing internal data structures, or adding a new consumer of it, should not require us to update the synchronization algorithm. Such coupling is error-prone. 

We have the same problem with propagating schema changes and we can solve it in the similar way. Introduce a per-node local hash of committed endpoint state info, calculated from endpoint states in the gossiper layer, and update it once all the listeners finish running. Then if we see that hashes match for all the nodes, it also means that all listeners have reacted.

I also think that there is no need to exchange anything more than a single hash. No need to send a map of keys and hashes. For debugging, one will have to do a full `nodetool gossipinfo` dump anyway to figure out exact difference.

Reply:
> There are two steps in the propagation. The first step is the exchange of information (endpoint states) which is done in the gossip layer. The second step is building/updating internal data structures from that, on all the shards. The synchronization algorithm should not know about higher layers which derive representation from that basic information. Changing internal data structures, or adding a new consumer of it, should not require us to update the synchronization algorithm. Such coupling is error-prone.
> 
> We have the same problem with propagating schema changes and we can solve it in the similar way. Introduce a per-node local hash of committed endpoint state info, calculated from endpoint states in the gossiper layer, and update it once all the listeners finish running. Then if we see that hashes match for all the nodes, it also means that all listeners have reacted.
> 

The ultimate goal is to make sure nodes in the cluster has the same view of the ring. Tokens and friends (like pending ranges) in the token_metadata are the authority of such view. No need to care about the two steps propagation. It does not matter. We only need to look at the final state of the tokens friends in the token_metadata. It is simple and has no dependency on how gossip run the listeners, whether synchronously or asynchronously. Plus, as I mentioned in the past, gossip is not the only source we update tokens in token_metadata. A node can update its own tokens directly into token_metadata without going through gossip. A node can update tokens of peer nodes from sytem.peers during boot up without going through gossip.

The check is sitting on a layer that consumes the queries from gossip and token_metadata. It is not an extension of gossip protocol. It  does not modify gossip protocol or add new application state to gossip. It does not add limitation on how listeners of gossip should work.

> I also think that there is no need to exchange anything more than a single hash. No need to send a map of keys and hashes. For debugging, one will have to do a full `nodetool gossipinfo` dump anyway to figure out exact difference.

What does a single hash buy us? What I can see is that it saves a few bytes in the response which is nothing. What do multiple  hashes give us? 1) When new items are added, mixed cluster old and new nodes can work with multiple hashes. Avi asked how we handle this in his review 2) When we decide some of the items are not needed, the receiver of the replies can choose to include specific hash or not in the comparison. It is more flexible. 3) nodetool gossipinfo does not give more detailed information like whether the pending ranges match or not. Nodetool gossipinfo does not show full and detailed tokens that a node owns. Even if it does, the user has to write a script and run it on all the nodes to figure out who is missing what, which we already have and we can get it in a single place if we use multiple hashes.

Please let me know why a single hash is better.

Reply:
> The check is sitting on a layer that consumes the queries from gossip and token_metadata. It is not an extension of gossip protocol. It does not modify gossip protocol or add new application state to gossip. It does not add limitation on how listeners of gossip should work.

And that is exactly why I believe it shouldn't be part of `gossiper`. It is in a higher layer that uses both `gossiper` and `token_metadata`. Maybe `storage_service`?
But `storage_service` is already this big monolith which we don't really know what it does (it does everything). So I'd probably use a free-standing function that takes gossiper and token_metadata as arguments, and call it from `storage_service` in the joining procedure.

However, we can do even better. `wait_for_gossip_with_query` consists of multiple things:
1. extracting information from our local gossiper (`get_gossip_query_token_status_response()`),
2. extracting information from other nodes (`send_gossip_query_token_status(...)`),
3. comparing our information to others' information.

But:
1. the algorithm doesn't/shouldn't need to know that the information about our local ring comes from `gossiper`; it could come from any propagation mechanism. It only cares about the information itself: i.e. the summary of my token ring. So it could get the summary as a parameter, not `gossiper`. If it needs to retrieve the summary continuously, it could take a function called `get_local_token_ring_summary` as a parameter.
2. even more important, the algorithm shouldn't need to know that the information about the other nodes' ring comes from *their* gossiper! It only cares about getting the information itself, i.e. the summaries of others' token rings. So again, it shouldn't be calling a function with `gossip` in its name. It could take a function called `get_remote_token_ring_summary` as a parameter.

If you want an example of this approach, I used it in cdc/generation.cc in `update_streams_description`. This function needed to retrieve the number of different token owners currently in our local token ring. But I didn't want it to be a part of `storage_service` or take `storage_service` as a parameter. So I gave it a `noncopyable_function<unsigned()> get_num_token_owners` parameter. It doesn't know where the function comes from, or how the number of token owners is retrieved - it just cares about getting the information. Then, in `storage_service` where I'm calling `update_streams_description`, I pass to it the following lambda: `[ss] { return ss->get_token_metadata().count_normal_token_owners(); }`. This way I have most of cdc-specific code separated from `storage_service`.

Reply:
Yes. I liked your idea to split it from gossip. 

We definitely can refactor in a way that 1) gossip is not part of the name 2) we have a new name for the new verb and the functions around it. It is more clear this way.

I will put the free functions in gossiper.cc for now, since it is the only consumer at the moment. We can move them to new place later if needed.